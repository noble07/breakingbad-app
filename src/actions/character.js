import { loadCharacters } from "services/loadCharacters";
import { types } from "types/types";

/**
 * 
 * Obtiene los personajes con las frases llamando al servicio loadCharacters
 * 
 */
export const startGettingCharacters = ({page, filter}) => {
  return async(dispatch, getState) => {

    const {filtered} = getState().char

    if (filter && filter !== '') {
      dispatch(filteredCharacters(true))
    } else if (filtered) {
      dispatch(filteredCharacters(false))
    }

    dispatch(startLoadingCharacters())
    const characters = await loadCharacters({page, filter})

    dispatch(setCharacters({characters}))
    dispatch(finishLoadingCharacters())
  }
}

/**
 * 
 * Agrega los personajes al estado global
 * 
 */

export const setCharacters = ({ characters }) => ({
  type: types.characterLoad,
  payload: characters
})

export const filteredCharacters = (filter) => ({
  type: types.characterFilter,
  payload: filter
})

export const startLoadingCharacters = () => ({ type: types.characterStartLoading })
export const finishLoadingCharacters = () => ({ type: types.characterFinishLoading })
