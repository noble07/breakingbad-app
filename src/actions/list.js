import { db } from "firebase/firebase-cofing"
import { loadLists } from "services/loadLists"
import { types } from "types/types"


export const startNewList = ({name}) => {
  return async(dispatch, getState) => {

    const {uid} = getState().auth

    try {

      const newList = {
        name
      }
      
      const doc = await db.collection(`${uid}/breakingbad/lists`).add(newList)

      dispatch(addNewList(doc.id, newList))

    } catch (error) {
      console.log(error)
    }

  }
}

export const addNewList = (id, list) => ({
  type: types.listAddNew,
  payload: {
    id,
    ...list
  }
})

export const startLoadingList = uid => {
  return async(dispatch) => {

    const lists = await loadLists(uid)

    dispatch(setLists(lists))

  }
}

export const startDeletingList = id => {
  return async(dispatch, getState) => {

    const {uid} = getState().auth

    await db.doc(`${uid}/breakingbad/lists/${id}`).delete()

    dispatch(deleteList(id))

  }
}

export const deleteList = id => ({
  type: types.ListDelete,
  payload: id
})

export const setLists = lists => ({
  type: types.listLoad,
  payload: lists
})

export const listLogout = () => ({
  type: types.listLogout
})