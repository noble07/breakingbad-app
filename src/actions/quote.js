import { db } from "firebase/firebase-cofing"

export const startRatingQuote = (quoteId, reaction) => {
  return async(dispatch, getState) => {

    const {uid} = getState().auth

    try {


      const quoteSnap = await db.collection(`${uid}/breakingbad/quotes`).where('idQuote', '==', quoteId).get()


      if (quoteSnap.empty){
      }

      quoteSnap.forEach(snapHijo => {
        console.log(snapHijo.data())
      });
      
    } catch (error) {
      console.log(error)
    }

  }
}