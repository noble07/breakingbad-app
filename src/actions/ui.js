import { types } from "types/types";

export const uiOpenModal = () => ({ type: types.uiOpenModal });

export const uiCloseModal = () => ({ type: types.uiCloseModal });

export const setError = (error) => ({ 
  type: types.uiSetError,
  payload: error
});

export const removeError = () => ({ type: types.uiRemoveError });

export const uiStartLoading = () => ({ type: types.uiStartLoading })

export const uiFinishLoading = () => ({ type: types.uiFinishLoading })