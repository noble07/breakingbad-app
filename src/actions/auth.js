import { firebase, googleAuthProvider } from 'firebase/firebase-cofing'
import { types } from "types/types";
import { uiCloseModal } from './ui';


export const startGoogleLogin = () => {
  return (dispatch) => {

      firebase.auth().signInWithPopup( googleAuthProvider )
          .then(({ user }) => {
              dispatch( login( user.uid, user.displayName ) );
              dispatch( uiCloseModal() )
          });

  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
      uid,
      displayName
  }
})

export const startLogout = () => {
  return async(dispatch) => {

      await firebase.auth().signOut();
      dispatch( logout() );

  }
}

export const logout = () => ({
  type: types.logout
})