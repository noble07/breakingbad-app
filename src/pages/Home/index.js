import { useEffect, useState } from "react";
import {
  Container,
  Dimmer,
  Divider,
  Image,
  Loader,
  Pagination,
  Segment } from 'semantic-ui-react'
import { useDispatch, useSelector } from "react-redux";

import FilterForm from 'components/Forms/FilterForm'
import CharactersList from 'components/Characters/CharactersList'
import LoginModal from 'components/LoginModal'

import { startGettingCharacters } from "actions/character";

import logo from 'assets/logo.png'

const Home = () => {

  const [page, setPage] = useState(1)
  const {loading} = useSelector(state => state.char)
  const dispatch = useDispatch()

  useEffect(() => {
    
    dispatch(startGettingCharacters({page}))

  }, [dispatch, page])

  const handlePageChange = ( e ,{activePage}) => {
    
    setPage(activePage)

  }
  

  return (
    <Container textAlign="center">
      <img className="logo-app" src={logo} alt="Logo" />
      <FilterForm />
      <Divider clearing />

      {
        (!loading) 
          ? <CharactersList />
          : (
            <Segment>
              <Dimmer active inverted>
                <Loader size="large">Loading</Loader>
              </Dimmer>

              <Image src="https://react.semantic-ui.com/images/wireframe/paragraph.png" />
            </Segment>
          )
      }



      <Pagination
        disabled={loading}
        style={{marginTop: 25, marginBottom: 25}}
        size="large"
        firstItem={null}
        lastItem={null}
        activePage={page}
        pointing
        secondary
        totalPages={4}
        onPageChange={handlePageChange}
      />

      <LoginModal />
    </Container> 
  )
}

export default Home
