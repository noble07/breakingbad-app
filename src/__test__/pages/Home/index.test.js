import { mount } from 'enzyme'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

import Home from 'pages/Home'
import { startGettingCharacters } from 'actions/character'

jest.mock('actions/character', () => ({
  startGettingCharacters: jest.fn()
}))


const middleware = [thunk]

const mockStore = configureStore(middleware)

const initialState = {
  char: {
    loading: false,
    characters: [
      {
        char_id: 123,
        name: 'Juan',
        birthday: '123123',
        img: 'https://sitio.com/prueba.png',
        nickname: 'Juank',
        quotes: [
          {
            quote_id: 123,
            quote: 'Prueba'
          }
        ]
      }
    ]
  },
  ui: {
    modalOpen: false
  }
}

describe('Pruebas en la pagina del Home', () => {
  
  test('debe de mostrarse correctamente', () => {

    const store = mockStore(initialState)
    store.dispatch = jest.fn
    
    const wrapper = mount(
      <Provider store={store}>
        <Home />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
    expect(startGettingCharacters).toHaveBeenCalled()

  })

  test('debe de mostrar mensaje de "No tiene frases el personaje"', () => {

    initialState.char.characters[0].quotes = []
    
    const store = mockStore(initialState)
    store.dispatch = jest.fn

    const wrapper = mount(
      <Provider store={store}>
        <Home />
      </Provider>
    )

    expect(wrapper.find('.warning').text()).toBe('El personaje no tiene frases')

  })
  

})
