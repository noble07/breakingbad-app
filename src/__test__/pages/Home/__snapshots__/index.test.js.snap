// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en la pagina del Home debe de mostrarse correctamente 1`] = `
<div
  className="ui center aligned container"
>
  <img
    alt="Logo"
    className="logo-app"
    src="logo.png"
  />
  <div
    className="ui centered vertically padded grid"
  >
    <div
      className="ui green compact raised center aligned segment"
    >
      <form
        className="ui form"
        onSubmit={[Function]}
      >
        <div
          className="inline field"
        >
          <label>
            Filtrar:
          </label>
          <div
            className="ui action input"
          >
            <input
              name="filter"
              onChange={[Function]}
              placeholder="Nombre"
              type="text"
            />
            <button
              className="ui button"
              onClick={[Function]}
              type="submit"
            >
              Buscar
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
  <div
    className="ui clearing divider"
  />
  <div
    className="ui two cards"
  >
    <div
      className="ui green card"
      onClick={[Function]}
      style={
        Object {
          "textAlign": "left",
        }
      }
    >
      <img
        className="image"
        src="https://sitio.com/prueba.png"
        style={
          Object {
            "height": 600,
            "objectFit": "cover",
            "objectPosition": "top",
            "width": "100%",
          }
        }
      />
      <div
        className="content"
      >
        <div
          className="header"
        >
          Juan
        </div>
        <div
          className="meta"
        >
          <span
            className="date"
          >
            Nacido el 
            123123
          </span>
        </div>
        <div
          className="description"
        >
          <div
            className="ui raised segments"
          >
            <div
              className="ui tiny segment"
            >
              Prueba
              <div
                className="ui fitted hidden divider"
              />
              Array [
                <div
                  className="ui right labeled button"
                  onClick={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <button
                    className="ui grey mini button"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <i
                      aria-hidden="true"
                      className="heart icon"
                      onClick={[Function]}
                    />
                  </button>
                </div>,
                <div
                  className="ui right labeled button"
                  onClick={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <button
                    className="ui grey mini button"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <i
                      aria-hidden="true"
                      className="star icon"
                      onClick={[Function]}
                    />
                  </button>
                </div>,
                <div
                  className="ui right labeled button"
                  onClick={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <button
                    className="ui grey mini button"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <i
                      aria-hidden="true"
                      className="fire icon"
                      onClick={[Function]}
                    />
                  </button>
                </div>,
              ]
              <button
                className="ui blue tiny button"
                onClick={[Function]}
              >
                <i
                  aria-hidden="true"
                  className="comment alternate outline icon"
                  onClick={[Function]}
                />
                Comentarios
              </button>
              <button
                className="ui brown small button"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <i
                  aria-hidden="true"
                  className="bookmark icon"
                  onClick={[Function]}
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="extra content"
      >
        Nickname: 
        Juank
      </div>
    </div>
  </div>
  <div
    aria-label="Pagination Navigation"
    className="ui large pagination pointing secondary menu"
    role="navigation"
    style={
      Object {
        "marginBottom": 25,
        "marginTop": 25,
      }
    }
  >
    <a
      aria-current={false}
      aria-disabled={false}
      aria-label="Previous item"
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="prevItem"
      value={1}
    >
      ⟨
    </a>
    <a
      aria-current={true}
      aria-disabled={false}
      className="active item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={1}
    >
      1
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={2}
    >
      2
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={3}
    >
      3
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={4}
    >
      4
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      aria-label="Next item"
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="nextItem"
      value={2}
    >
      ⟩
    </a>
  </div>
</div>
`;
